grains: add test template

smalley opened the pull request:
Add test template for grains which adds all square() tests and the method to generate the only existing total() test. Closes #1952.

smalley commented:
I'm not certain this is complete. The majority of tests are included here, although the description generated names based on the canonical data do not match the original test case names. 

Additionally it does not seem like that canonical data currently includes the tests for total() although it includes a single (what seems like incorrect) test of total(). I had separately opened exercism/problem-specifications#1613 which attempts to correct at least this single existing total test. I'm not sure what the best path forward here would be.

yawpitch commented:
The `total` function should never have had a parameter, that's a mistake. It should have one test, no input, and it should return the sum of all 64 squares.

Please update the tests, the example.py, and the grains.py slug file to reflect that.

yawpitch commented:
While you're updating the example.py maybe remove the **check_square_input** helper function and incorporate its functionality into **square**.

smalley commented:
Updated the template to use the error_case filter correctly, changed the example.py and slug to show total() taking no parameter. @yawpitch I remove the helper function from total but would you prefer total return the hardcoded answer for 64 or still calculate the answer using squar() (which also calls the test).

yawpitch commented:
> Updated the template to use the error_case filter correctly, changed the example.py and slug to show total() taking no parameter. @yawpitch I remove the helper function from total but would you prefer total return the hardcoded answer for 64 or still calculate the answer using squar() (which also calls the test).

Rather than hardcoded I’d do `return 2 ** 64 - 1`

> Updated the template to use the error_case filter correctly, changed the example.py and slug to show total() taking no parameter. @yawpitch I remove the helper function from total but would you prefer total return the hardcoded answer for 64 or still calculate the answer using squar() (which also calls the test).



> Updated the template to use the error_case filter correctly, changed the example.py and slug to show total() taking no parameter. @yawpitch I remove the helper function from total but would you prefer total return the hardcoded answer for 64 or still calculate the answer using squar() (which also calls the test).

I’d tend to do `return (2 ** 64) - 1` rather than the constant, just to demonstrate the nice symmetry with `return 2 ** (64 - 1)`.

smalley commented:
@yawpitch Okay, made one last change to use 2 ** (64-1) as the style for total() in example.py

smalley commented on exercises/grains/.meta/template.j2:
```diff
@@ -0,0 +1,31 @@
+        {%-endif%}
+    {% endfor %}
+
+
```
Okay I think I have the spacing in template.j2 resolved. Does this need to be rebased?

cmccandless commented on exercises/grains/.meta/template.j2:
```diff
@@ -0,0 +1,29 @@
+          with self.assertRaisesWithMessage(ValueError):
+                {{ case["property"] | to_snake }}({{ input["square"] }})
+    {%- else%}
+    def test_{{case["property"] | to_snake}}_{{ case["description"] | to_snake }}(self):
```
Since we no longer test for parameters to `total()`, I think this can change as follows:
```suggestion
    def test_{{ case["description"] | to_snake }}(self):
```
...which should allow the template to be simplified a bit more.

smalley replied:
Made the swap to the function definition line, unified the cases so that the function definition occurs outside the conditional.

cmccandless commented:
> Honestly I just wasn't sure what the best practice is for contributing with PR changes.

I'm not sure if there's a written-down best practice, but my *personal* preference is to leave the individual commits in your branch. This allows for two things:
- The entire change process is visible for the record
- When I get a notification for changes made to a PR, GitHub links to the diff page between the last commit I reviewed and the latest changes. However, if the last commit I reviewed no longer exists (wiped out by `push --force`), the page fails to load at all. I no longer have any way to see what has changed since I last reviewed, and I have to rely on memory and out-of-context review comments to see if my requests have been completely addressed.

> I assume that before a fix/feature is merged to master you would want it as a single discrete commit for small things like this but that I wouldn't necessarily have an opportunity to do a rebase and squash if each requested change was an additional branch commit.

I (as a maintainer) have the option to do `merge --squash` for a PR, and that is my default. This accomplishes the single commit goal, without squashing the history of the PR.
